File: internal/constraints/string_builtins_test.go
package constraints

import (
	"testing"
	"time"

	"github.com/opentalon/tln-language/internal/ast"
	"github.com/opentalon/tln-language/internal/lexer"
	"github.com/opentalon/tln-language/internal/parser"
)

// evalStr parses a single expression (as the left side of a throwaway detect
// selector's compare) and evaluates it against the given row.
func evalStr(t *testing.T, expr string, row map[string]any) any {
	t.Helper()
	src := `detect "T" { for records where ` + expr + ` == "SENTINEL" }`
	tokens, ld := lexer.Lex("t.tln", src)
	if ld.HasErrors() {
		t.Fatalf("lex %q: %v", expr, ld)
	}
	prog, pd := parser.Parse("t.tln", tokens)
	if pd.HasErrors() {
		t.Fatalf("parse %q: %v", expr, pd)
	}
	det, ok := prog.Blocks[0].(*ast.DetectBlock)
	if !ok || len(det.Selector.Conditions) == 0 {
		t.Fatalf("parse %q: no selector condition", expr)
	}
	cmp, ok := det.Selector.Conditions[0].(*ast.CompareCondition)
	if !ok {
		t.Fatalf("parse %q: selector condition is %T, want compare", expr, det.Selector.Conditions[0])
	}
	v, err := EvalExpr(cmp.Left, row, time.Unix(0, 0).UTC())
	if err != nil {
		t.Fatalf("eval %q: %v", expr, err)
	}
	return v
}

func TestStringBuiltins(t *testing.T) {
	row := map[string]any{
		"name": "  Broken Drill  ",
		"sku":  "AB1234",
		"vin":  "1ftabc",
		"n":    float64(3),
	}
	cases := []struct {
		expr string
		want any
	}{
		{`upper(attr "vin")`, "1FTABC"},
		{`lower(attr "sku")`, "ab1234"},
		{`trim(attr "name")`, "Broken Drill"},
		{`length(attr "sku")`, float64(6)},
		{`substring(attr "sku", 0, 2)`, "AB"},
		{`substring(attr "sku", 2)`, "1234"},
		{`replace(attr "sku", "AB", "XY")`, "XY1234"},
		{`concat("v-", attr "vin", "-", attr "n")`, "v-1ftabc-3"}, // number stringifies without .0
		{`join(split(attr "sku", "1"), "_")`, "AB_234"},           // split → ["AB","234"], join → "AB_234"
		{`upper(substring(attr "vin", 0, 3))`, "1FT"},             // nested
	}
	for _, c := range cases {
		got := evalStr(t, c.expr, row)
		if got != c.want {
			t.Errorf("%s = %#v, want %#v", c.expr, got, c.want)
		}
	}
}

File: packages/runtime/__tests__/string_builtins.test.ts
import { describe, it, expect } from "vitest"
import { resolveExpr, type EvalContext } from "../src/evaluator"
import { lex } from "../src/lexer"
import { parse } from "../src/parser"
import { TlnStore } from "../src/store"

// evalExprString parses `<expr>` as the left side of a rule condition and
// evaluates it against the given state.
function evalExprString(expr: string, state: Record<string, any> = {}): any {
  const prog = parse(lex(`rule "T" { when ${expr} == "SENTINEL" do noop }`))
  const when = prog.rules[0].when as any
  const ctx: EvalContext = {
    state: new Map(Object.entries(state)),
    defines: new Map(),
    changedPaths: new Set(),
  }
  return resolveExpr(when.left, ctx)
}

describe("string builtins — evaluation", () => {
  const row = { vin: "1ftabc", sku: "AB1234", name: "  Broken  " }
  it("upper / lower / trim / length", () => {
    expect(evalExprString(`upper(vin)`, row)).toBe("1FTABC")
    expect(evalExprString(`lower(sku)`, row)).toBe("ab1234")
    expect(evalExprString(`trim(name)`, row)).toBe("Broken")
    expect(evalExprString(`length(sku)`, row)).toBe(6)
  })
  it("substring (2- and 3-arg) and nesting", () => {
    expect(evalExprString(`substring(sku, 0, 2)`, row)).toBe("AB")
    expect(evalExprString(`substring(sku, 2)`, row)).toBe("1234")
    expect(evalExprString(`upper(substring(vin, 0, 3))`, row)).toBe("1FT")
  })
  it("replace / concat / split / join", () => {
    expect(evalExprString(`replace(sku, "AB", "XY")`, row)).toBe("XY1234")
    expect(evalExprString(`concat("v-", vin, "-", length(sku))`, row)).toBe("v-1ftabc-6")
    expect(evalExprString(`join(split(sku, "1"), "_")`, row)).toBe("AB_234")
  })
})

describe("string builtins — in a rule action", () => {
  it("resolves a string function as an action argument", () => {
    const store = new TlnStore()
    const fired: any[] = []
    store.set("recall_code", "rb7")
    store.loadRules(`
      rule "Recall" {
        when "recall_code" changes
        do open_recall upper(recall_code)
      }
    `)
    store.subscribe("*", (a) => fired.push(...a))
    store.set("recall_code", "rb7")
    const call = fired.find((a) => a.verb === "open_recall")
    expect(call?.args[0]).toBe("RB7")
  })
})
